### PR TITLE
dpdk_telemetry plugin: add plugin for DPDK metrics via DPDK Telemetry library

### PR DESCRIPTION
**Pull request by reshmapattan**
ChangeLog: dpdk_telemetry plugin: New plugin to fetch DPDK metrics.

This patch introduces a new plugin for collectd, which consumes DPDK metrics
via the dpdk_telemetry library. The collectd plugin here provides an
easy way to use the DPDK telemetry API to query ethernet device metrics.

The collectd plugin retrieves metrics from a DPDK packet forwarding
application by sending a JSON formatted message via a UNIX domain
socket. The DPDK telemetry component will respond with a JSON formatted
reply, delivering the requested metrics. The dpdk_telemetry collectd
plugin parses the JSON data, and publishes the metric values to collectd
for further use.

This plugin has a dependency on the DPDK Telemetry library, as it must be
"in sync" with the DPDK Telemetry implementation.

Signed-off-by: Emma Kenny <emma.kenny@intel.com>
Signed-off-by: Brian Archbold <brian.archbold@intel.com>
Signed-off-by: Reshma Pattan <reshma.pattan@intel.com>

### CODE REVIEW

**Comment by reshmapattan** [comment]
Hi ,

I am a new contributor to the collectd and I have pushed new plugin and the build seems failing 
https://travis-ci.org/collectd/collectd/jobs/583765756
Not clear with the reason of the build failure,  please let me know what is the issue and how to fix it.
Appreciate the help.

Thanks,
Reshma

**Comment by rpv-tomsk** [comment]
This is typical for Collectd infrastructure to have continuosly failing CI checks.

Nobody cares.

>I am a new contributor to the collectd

Don't think things happen fast here. Some guys waits for years before his PR gets merged (Personally, I have been waiting for months).

This is a harsh truth, sorry for bad news.

**Comment by sunkuranganath** [comment]
Thanks @reshmapattan for your contributions. Looking in to the CI issues now.

**Comment by rpv-tomsk** [comment]
@sunkuranganath, you is responsible person for Collectd CI? 
You have appropriate permissions and technical accesses? When do you received them?

If you is responsible for Collectd CI, will you solve issue wide, or you will help only Intel people, leaving other away (example: #3264)?

**Comment by rpv-tomsk** [comment]
From my POV, this plugin implementation is strange and needs to be reworked.
There should be no `while(1)` in `read_callback` and `init_callback`. 
Plugin should not fail if there is no connection on `init`. That is not critical state, connection it can be restored later. There should not be race conditions on daemons starting.
And many other remarks, which I don't want to waste time.

**Comment by rpv-tomsk** [comment]
But `Trusted Maintainers` can merge this as is, of course.

**Comment by rpv-tomsk** [comment]
>Some sort of response is better than nothing, especially in open source community.
That's why my #3159 left opened w/o any comments for 4 months. Ok....

**Comment by rpv-tomsk** [comment]
>Looking in to the CI issues now.

So, you simply can't do anything with this.
Do not reassure others and do not create illusions about community health. 
No such problems has been solved over last year, at least.

**Comment by reshmapattan** [comment]
> From my POV, this plugin implementation is strange and needs to be reworked.
> There should be no `while(1)` in `read_callback` and `init_callback`.
> Plugin should not fail if there is no connection on `init`. That is not critical state, connection it can be restored later. There should not be race conditions on daemons starting.
> And many other remarks, which I don't want to waste time.

I have pushed the new changes by removing the `while(1)` loops.  Please provide your feedback.

**Comment by mrunge** [comment]
The CI job failed due to network (or ubuntu mirror issues):
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/s/systemd/udev_229-4ubuntu21.22_amd64.deb  Could not connect to apt.cache.travis-ci.com:80 (34.96.81.152), connection timed out

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/s/systemd/libudev1_229-4ubuntu21.22_amd64.deb  Unable to connect to apt.cache.travis-ci.com:http:

E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/liba/libatasmart/libatasmart4_0.19-3_amd64.deb  Unable to connect to apt.cache.travis-ci.com:http:

**Comment by mrunge** [comment]
Would it be possible to use something like yajl instead of another json library?

**Comment by kkepka** [comment]
[yajl](https://github.com/lloyd/yajl) seems to be not active (last commit is from 2015), that's why [jansson ](https://github.com/akheron/jansson) has been chosen instead.

**Comment by mrunge** [comment]
> [yajl](https://github.com/lloyd/yajl) seems to be not active (last commit is from 2015), that's why [jansson ](https://github.com/akheron/jansson) has been chosen instead.

Oh, I didn't realize that. Good point!

**Comment by mrunge** [comment]
So, CI failed due to networking issues, not because of the submission itself.

File: configure.ac [hunk 1]
```diff
@@ -6802,6 +6852,7 @@ AC_PLUGIN([dns],                 [$with_libpcap],             [DNS traffic analy
 AC_PLUGIN([dpdkevents],          [$plugin_dpdkevents],        [Events from DPDK])
 AC_PLUGIN([dpdkstat],            [$plugin_dpdkstat],          [Stats from DPDK])
 AC_PLUGIN([drbd],                [$plugin_drbd],              [DRBD statistics])
+AC_PLUGIN([dpdk_telemetry],      [$plugin_dpdk_telemetry],    [Metrics from DPDK Telemetry])
```
Comment (rpv-tomsk) [inline configure.ac]: Please, keep list sorted.

Comment (reshmapattan) [reply configure.ac]: Addressed and available in latest commit.

Comment (reshmapattan) [reply configure.ac]: @rpv-tomsk , Can you please approve the pull request , if no  more comments.

Comment (rpv-tomsk) [reply configure.ac]: Sorry, I'm `untrusted contributor` and have no such permission.

File: src/dpdk_telemetry.c [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+          ERROR(PLUGIN_NAME ": Parsing failed");
+      }
+    }
+  }
+  return 0;
+}
+
+static int dpdk_telemetry_init(void) {
+
+  INFO(PLUGIN_NAME ": %s:%d", __FUNCTION__, __LINE__);
```
[the remark this answers is not in the document]
Comment (reshmapattan) [reply src/dpdk_telemetry.c]: Hi,

Tanks for the approval.  

IMO, Information on plugin status would be handy to know plugin running status, so would prefer to keep  this as is and in other places.

[the remark this answers is not in the document]
Comment (anaudx) [reply src/dpdk_telemetry.c]: ok

File: configure.ac [hunk 3]
```diff
@@ -4541,6 +4541,51 @@ AC_SUBST([BUILD_WITH_LIBPQOS_LDFLAGS])
...
+
+  AC_CHECK_HEADERS([jansson.h],
+    [with_libjansson="yes"],
+    [with_libjansson="no (jansson.h not found)"]
+  )
+ CPPFLAGS="$SAVE_CPPFLAGS"
+fi
+if test "x$with_libjansson" = "xyes"; then
+  SAVE_LDFLAGS="$LDFLAGS"
+  LDFLAGS="$LDFLAGS $with_jansson_ldflags"
```
[the remark this answers is not in the document]
Comment (reshmapattan) [reply configure.ac]: ohh yes, also in other 2 lines. Will push the changes.

**Comment by reshmapattan** [comment]
@mrunge : can this be mergered to master now.

**Comment by mrunge** [comment]
I am unfortunately utterly busy with other things currently. This will have to wait (or until someone else approves this PR).

Self approving won't help you. In any case, *you* could help by reviewing other PRs.

**Comment by reshmapattan** [comment]
> I am unfortunately utterly busy with other things currently. This will have to wait (or until someone else approves this PR).
> 
> Self approving won't help you. In any case, _you_ could help by reviewing other PRs.
Ok , if you are busy I can wait.  In case if you missed it  is this already aprroved by mkobyli .

**Comment by kkepka** [comment]
I believe @reshmapattan just wanted to friendly ping / bump this up here, as 5.10 is out and window for new features for 5.11 is open now.
I wouldn't see it as self-approved, more like feedback for maintainers and others. Sorry if you see it that way, maybe in future such feedback could be provided via comment only (without approval checkbox?)

**Comment by sunkuranganath** [comment]
Given there arent any more comments/review, i hope we can approve this to make in to 5.11 release.

**Comment by mrunge** [comment]
> Given there arent any more comments/review, i hope we can approve this to make in to 5.11 release.

There hasn't been any serious reviews for now....

**Pull request merged**